src/dbnode/namespace/namespace_runtime_options.go

```diff
@@ -0,0 +1,252 @@
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"sync"
+
+	xclose "github.com/m3db/m3/src/x/close"
+	"github.com/m3db/m3/src/x/watch"
+)
+
+const (
+	defaultWriteIndexingPerCPUConcurrency = 0.75
+	defaultFlushIndexingPerCPUConcurrency = 0.25
+)
+
+// RuntimeOptions is a set of runtime options that can
+// be set per namespace.
+type RuntimeOptions interface {
+	// IsDefaults returns whether the runtime options are purely defaults
+	// with no values explicitly set.
+	IsDefaults() bool
+
+	// Equal will return whether it's equal to another runtime options.
+	Equal(other RuntimeOptions) bool
+
+	// SetWriteIndexingPerCPUConcurrency sets the write
+	// indexing per CPU concurrency.
+	SetWriteIndexingPerCPUConcurrency(value *float64) RuntimeOptions
+
+	// WriteIndexingPerCPUConcurrency returns the write
+	// indexing per CPU concurrency.
+	WriteIndexingPerCPUConcurrency() *float64
+
+	// WriteIndexingPerCPUConcurrencyOrDefault returns the write
+	// indexing per CPU concurrency or default.
+	WriteIndexingPerCPUConcurrencyOrDefault() float64
+
+	// SetFlushIndexingPerCPUConcurrency sets the flush
+	// indexing per CPU concurrency.
+	SetFlushIndexingPerCPUConcurrency(value *float64) RuntimeOptions
+
+	// FlushIndexingPerCPUConcurrency returns the flush
+	// indexing per CPU concurrency.
+	FlushIndexingPerCPUConcurrency() *float64
+
+	// FlushIndexingPerCPUConcurrencyOrDefault returns the flush
+	// indexing per CPU concurrency.
+	FlushIndexingPerCPUConcurrencyOrDefault() float64
+}
+
+// RuntimeOptionsManagerRegistry is a registry of runtime options managers.
+type RuntimeOptionsManagerRegistry interface {
+	// RuntimeOptionsManager returns a namespace runtime options manager
+	// for the given namespace.
+	RuntimeOptionsManager(namespace string) RuntimeOptionsManager
+
+	// Close closes the watcher and all descendent watches.
+	Close()
+}
+
+// RuntimeOptionsManager is a runtime options manager.
+type RuntimeOptionsManager interface {
+	// Update updates the current runtime options.
+	Update(value RuntimeOptions) error
+
+	// Get returns the current values.
+	Get() RuntimeOptions
+
+	// RegisterListener registers a listener for updates to runtime options,
+	// it will synchronously call back the listener when this method is called
+	// to deliver the current set of runtime options.
+	RegisterListener(l RuntimeOptionsListener) xclose.SimpleCloser
+
+	// Close closes the watcher and all descendent watches.
+	Close()
+}
+
+// RuntimeOptionsListener listens for updates to runtime options.
+type RuntimeOptionsListener interface {
+	// SetNamespaceRuntimeOptions is called when the listener is registered
+	// and when any updates occurred passing the new runtime options.
+	SetNamespaceRuntimeOptions(value RuntimeOptions)
+}
+
+// runtimeOptions should always use pointer value types for it's options
+// and provide a "ValueOrDefault()" method so that we can be sure whether
+// the options are all defaults or not with the "AllDefaults" method.
```

Comment:
Is this preferable to having `writeIndexingPerCPUConcurrency` be a
```golang
struct {
 val float64
 isDefault bool
}
```
?

Reply:
For this case it was preferable since we needed to know later whether it was explicitly set or not (for when we do a namespace update with the API).